### DISCUSSION TITLE
Animating a glyphed point cloud

### DISCUSSION THREAD

**Discussion by apcode** [pyvista/pyvista#4317] [Q&A]
**Describe the bug, what's wrong, and what you expect:**
I am trying to animate a point cloud of points that move over time. Trying to use the plotter with points rendered as spheres in a window. I do not want to do this off-screen as a gif or mov.

I am trying to use the plotter update_coordinates with interactive_update=True. As I understand it I should be able to change the points and have the changes rendered on screen with this approach.

`plotter.show(window_size=(1024, 1024), interactive_update=True)`

then `plotter.update_coordinates(points, mesh=glyphs, render=True)`
and `plotter.update()`

The script crashes on the call to update_coordinates.

I would like help to get this working, or if there is a preferred way to animate a pyvista plotter render on screen can you advise how to achieve this.

** To Reproduce **
```py
import pyvista as pv
import numpy as np
import random
import time

# Initialize particle data
num_particles = 100
positions = np.random.rand(num_particles, 3)
velocities = (np.random.rand(num_particles, 3) - 0.5) * 0.1

# Create a PyVista mesh for the particles
points = pv.PolyData(positions)
points["size"] = np.random.rand(num_particles) * 0.05 + 0.05

sphere = pv.Sphere(radius=0.3)
glyphs = points.glyph(geom=sphere, scale="size", factor=1, rng=(0.05, 0.1))

# Create a PyVista plotter
plotter = pv.Plotter()

# Add particles to the plotter
plotter.add_mesh(glyphs, show_scalar_bar=False, name="points")

# Set up a callback function for updating particle positions
def update_points():
    global positions, velocities

    # Update positions based on velocities
    positions += velocities

    # Update velocities based on particle interactions (you can replace this with your custom logic)
    velocities += (np.random.rand(num_particles, 3) - 0.5) * 0.01

    return positions


plotter.show(window_size=(1024, 1024), interactive_update=True)

for n in range(10):
    print(n)
    points = update_points()
    plotter.update_coordinates(points, mesh=glyphs, render=True)
    plotter.update()
    time.sleep(1)
```

** Result **
```
python(98546,0x1e6599b40) malloc: Heap corruption detected, free list is damaged at 0x600003e16e60
*** Incorrect guard value: 72340172838076673
python(98546,0x1e6599b40) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

** System Information **

--------------------------------------------------------------------------------
  Date: Sun Apr 23 13:59:15 2023 BST

                OS : Darwin
            CPU(s) : 10
           Machine : arm64
      Architecture : 64bit
               RAM : 32.0 GiB
       Environment : Python
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M1 Max
       GPU Version : 4.1 Metal - 83.1

  Python 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)
  [Clang 13.0.1 ]

           pyvista : 0.34.1
               vtk : 9.1.0
             numpy : 1.23.3
           imageio : 2.10.3
           appdirs : 1.4.4
            scooby : 0.5.12
        matplotlib : 3.4.3
           IPython : 8.5.0
             scipy : 1.9.1
              tqdm : 4.64.1
            meshio : 5.3.4
--------------------------------------------------------------------------------

**Comment by banesullivan** [comment]
Few things are going awry here. The primary issue is that you have a pipeline. The glyph's you are tying to update are computed from the `glyph()` filter. You will need to re-run this filter each time you update the positions of the `points` mesh.

I recommend avoiding the `update_coordinates` API and update the meshes directly. This will get tricky with the glyphs as you may need to add/re-add them to the plotter on each iteration using a unique `name` argument

HTH

**Reply by AlkisPis** [reply]
> update the meshes directly

How is updating the meshes directly done? That is, what must the calls to `update_coordinates()` and `update_scalars()` be replaced with exactly?

**Reply by akaszynski** [reply]
> How is updating the meshes directly done? That is, what must the calls to update_coordinates() and update_scalars() be replaced with exactly?

Please see https://github.com/pyvista/pyvista/discussions/4317#discussioncomment-5767802

**Reply by AlkisPis** [reply]
> > How is updating the meshes directly done? That is, what must the calls to update_coordinates() and update_scalars() be replaced with exactly?
> 
> Please see [#4317 (comment)](https://github.com/pyvista/pyvista/discussions/4317#discussioncomment-5767802)

Thanks, but this doesn't help at all. I have asked for an **"exact" solution**, meaning based on the code that @apcode is given. The example you are referring to --even if it solves the problem --I couldn't know-- has nothing to do with that.

**Reply by akaszynski** [reply]
Here's two examples, one with and without glyphs:

#### Without Glyphs
```py
"""Glyph-free example.

Particles are rendered uniformly.

"""

import pyvista as pv
import numpy as np

velocity_factor = 1e-3

# Initialize particle data
num_particles = 100

# we're using a list here to allow updating in-place without using a global
velocities = [(np.random.rand(num_particles, 3) - 0.5) * velocity_factor]

# Create a PyVista mesh for the particles
particles = pv.PolyData(np.random.rand(num_particles, 3))

# Create a PyVista plotter
pl = pv.Plotter()
actor = pl.add_mesh(
    particles,
    show_scalar_bar=False,
    name="points",
    render_points_as_spheres=True,
    point_size=10,
)
pl.open_movie("glyph-free.mp4", framerate=60)


def update_points():
    # Update positions based on velocities
    particles.points += velocities[0]

    # Update velocities based on particle interactions (you can replace this
    # with your custom logic)
    velocities[0] += (np.random.rand(num_particles, 3) - 0.5) * velocity_factor


pl.show(interactive_update=True)

for n in range(500):
    update_points()
    # pl.render()  # unnecessary if we're writing a frame
    pl.write_frame()

pl.close()

```

https://github.com/user-attachments/assets/8853fbd1-fd70-44c9-ae5a-402f61d987a7


#### With Glyphs

```py
"""Glyph-free example.

Particles are rendered uniformly.

"""

import pyvista as pv
import numpy as np

velocity_factor = 1e-3
size_factor = 0.05

# Initialize particle data
num_particles = 100

# we're using a list here to allow updating in-place without using a global
velocities = [(np.random.rand(num_particles, 3) - 0.5) * velocity_factor]

# Create a PyVista mesh for the particles
particles = pv.PolyData(np.random.rand(num_particles, 3))

init_size = np.random.rand(num_particles) * size_factor + size_factor
final_size = np.random.rand(num_particles) * size_factor + size_factor * 2

particles["size"] = init_size

sphere = pv.Sphere(radius=0.3)
glyphs = particles.glyph(
    geom=sphere, scale="size", factor=1, rng=(size_factor, size_factor * 2)
)

# Create a PyVista plotter
pl = pv.Plotter()

# Add particles to the plotter
actor = pl.add_mesh(
    glyphs, show_scalar_bar=False, name="points", cmap="bwr", smooth_shading=True
)
pl.open_movie("particles-glyph.mp4", framerate=60)

n_frames = 500


def update_points(frame_n):
    # Update positions based on velocities
    particles.points += velocities[0]

    # allow the particles to get bigger over time
    ratio = frame_n / n_frames
    new_size = init_size * (1 - ratio) + final_size * ratio
    print(new_size[0])
    particles["size"] = new_size

    # Update velocities based on particle interactions (you can replace this
    # with your custom logic)
    velocities[0] += (np.random.rand(num_particles, 3) - 0.5) * velocity_factor

    # we have to overwrite the glyph dataset since we're changing the underlying mesh
    actor.mapper.dataset = particles.glyph(
        geom=sphere, scale="size", factor=1, rng=(size_factor, 0.1)
    )


pl.show(interactive_update=True)

for n in range(n_frames):
    update_points(n)
    # pl.render()  # not needed when writing a frame
    pl.write_frame()

pl.close()
```


https://github.com/user-attachments/assets/b304a3c9-762e-4010-b004-905fab4bca0b

**Comment by akaszynski** [comment]
Agree with @banesullivan's assessment. Here's the code to do this:


https://user-images.githubusercontent.com/11981631/235374452-043a3ede-1b3c-4df8-b889-d2eb7a5247cb.mp4


```py
from tqdm import tqdm
import pyvista as pv
import numpy as np
import random
import time

pv.set_plot_theme('document')

# Initialize particle data
num_particles = 100
positions = np.random.rand(num_particles, 3)

# Create a PyVista mesh for the particles
poly = pv.PolyData(positions)
poly['velocities'] = (np.random.rand(num_particles, 3) - 0.5) * 0.001
poly['size'] = np.random.rand(num_particles) * 0.05 + 0.05

sphere = pv.Sphere(radius=0.3)

# Create a PyVista plotter
pl = pv.Plotter(off_screen=True)

# Set up a callback function for updating particle positions
def gen_glyphs():
    """Update point positions and return glyphs of those particles."""
    poly.points += poly['velocities']

    # Update velocities based on particle interactions (insert logic here...)
    poly['velocities'] += (np.random.rand(num_particles, 3) - 0.5) * 0.01
    poly['size'] += np.random.rand(num_particles) * 0.05 - 0.025
    return poly.glyph(geom=sphere, scale='size', factor=1, rng=(0.05, 0.1), orient=False)

# add the mesh, open the movie, and write the first frame
pl.add_mesh(gen_glyphs(), smooth_shading=True, name='points', show_scalar_bar=False)
pl.enable_anti_aliasing('ssaa')
pl.open_movie('/tmp/sample.mp4', quality=10)
pl.write_frame()

# write the remaining frames
for n in tqdm(range(260)):
    pl.remove_actor('points')  # Remove the previous actor
    pl.add_mesh(gen_glyphs(), smooth_shading=True, name='points', show_scalar_bar=False)
    pl.write_frame()

pl.close()
```
